### PR TITLE
[java] Refactored god class rule

### PR DESCRIPTION
**Pull request by oowekyala**
This is the new god class rule

The metrics can be improved individually later, especially ATFD, which has some known limitations (I'll have no time to do that in a few days)

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java [hunk 1]
```diff
@@ -25,6 +25,48 @@ private StringUtil() {
...
+     * @return A formatted string
+     *
+     * @throws IllegalArgumentException if the double to format is not between 0 and 1
+     */
+    public static String percentageString(double val, int numDecimals, boolean truncateInt) {
+        if (val < 0 || val > 1) {
+            throw new IllegalArgumentException("Expected a number between 0 and 1");
+        }
+
+        return truncateDouble(100 * val, numDecimals, truncateInt) + "%";
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java]: I think, you could use `String.format(...)` here:

    return String.format("%." + numDecimals + "f%%", 100 * val);

I think, this is without the truncate-to-Int functionality.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/TccMetric.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+     * @return The number of pairs
+     */
+    private int numMethodsRelatedByAttributeAccess(Map<String, Set<String>> usagesByMethod) {
+        List<String> methods = new ArrayList<>(usagesByMethod.keySet());
+        int methodCount = methods.size();
+        int pairs = 0;
+
+        if (methodCount > 1) {
+            for (int i = 0; i < methodCount; i++) {
+                for (int j = i + 1; j < methodCount; j++) {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/TccMetric.java]: That would compare in the last round the same method - i == j.
I think the condition for the out loop over i should be "i < methodCount - 1"?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        } else {
+            result = true;
+        }
+
+        return result;
+    }
+
+
+    private String getNameImage(ASTPrimaryExpression node) {
+        ASTPrimaryPrefix prefix = node.getFirstDescendantOfType(ASTPrimaryPrefix.class);
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java]: Could there be a case, where there is no PrimaryPrefix? Do we need to check for null?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        String image = null;
+        if (name != null) {
+            image = name.getImage();
+        }
+        return image;
+    }
+
+
+    private String getMethodOrAttributeName(ASTPrimaryExpression node) {
+        ASTPrimaryPrefix prefix = node.getFirstDescendantOfType(ASTPrimaryPrefix.class);
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java]: Could there be a case, where there is no PrimaryPrefix? Do we need to check for null?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/GodClassRule.java [hunk 5]
```diff
@@ -4,366 +4,56 @@
...
-     * @return
+     * Few means between 2 and 5. See: Lanza. Object-Oriented Metrics in Practice. Page 18.
      */
-    private double calculateTotalMethodPairs() {
-        int methodCount = methodAttributeAccess.size();
-        int n = methodCount - 1;
-        double totalMethodPairs = n * (n + 1) / 2.0;
-        return totalMethodPairs;
-    }
+    private static final int FEW_ATFD_THRESHOLD = 6;
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/GodClassRule.java]: FEW was before 5. Any reason why you changed it to 6?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/GodClassRule.java [hunk 6]
```diff
@@ -4,366 +4,56 @@
...
-                }
-            }
-        }
-
-        return super.visit(node, data);
-    }
+    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
+        int wmc = (int) JavaMetrics.get(JavaClassMetricKey.WMC, node);
+        double tcc = JavaMetrics.get(JavaClassMetricKey.TCC, node);
+        int atfd = (int) JavaMetrics.get(JavaClassMetricKey.ATFD, node);
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/GodClassRule.java]: So simple! I like it!

File: pmd-java/src/main/resources/rulesets/java/design.xml [hunk 7]
```diff
@@ -1856,7 +1856,8 @@ public class HelloWorldBean {
   <rule name="GodClass"
         language="java"
         since="5.0"
-        message="Possible God class"
+        message="Possible God Class (WMC={0}, ATFD={2}, TCC={1})"
```
Comment (adangel) [inline pmd-java/src/main/resources/rulesets/java/design.xml]: Nice idea to ouput the metric values

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/AtfdTest.xml [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            <message>'.StatementAndBraceFinder#tryToLog(String, NodeType, Node)' has value 4.</message>
+            <message>'.StatementAndBraceFinder#tryToLog(NodeType, Node)' has value 0.</message>
+            <message>'.StatementAndBraceFinder#visit(ASTStatementExpression, Object)' has value 4.</message>
+            <message>'.StatementAndBraceFinder#visit(ASTVariableDeclarator, Object)' has value 4.</message>
+            <message>'.StatementAndBraceFinder#visit(ASTExpression, Object)' has value 18.</message>
+        </expected-messages>
+        <code-ref id="full-example"/>
+    </test-code>
+
+    <test-code>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/AtfdTest.xml]: you can add the attribute `regressionTest="true`", then this test case will be ignored.... You can then expect the correct message.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/TccMetric.java (hunk 2 above)
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/TccMetric.java]: Hum no, in the last round,  i = methodCount - 1, so j = i + 1 = methodCount, thus `j < methodCount` doesn't hold and the inner for loop is skipped. You wrote this ;)

Changing the outer condition might be clearer though, will do

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java (hunk 3 above)
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java]: I've never seen that case, not sure if that's possible... Maybe @jsotuyod would know?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/GodClassRule.java (hunk 5 above)
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/GodClassRule.java]: Err no valid reason. I guess it's not uncommon to use getters or setters of the objects you manipulate, and a total of 5 in a class seemed a bit low (to me). I don't know if it makes a difference really, I can put back 5 so as not to change the behaviour too much

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/AtfdTest.xml (hunk 8 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/AtfdTest.xml]: Nice, thanks :)

File: pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java (hunk 1 above)
Comment (oowekyala) [inline pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java]: That's nice indeed! We'll have to override the locale though, the decimal separator is locale-specific (for me it's a comma). Tests that pass on my machine could fail on other machines.

Truncate to int is only really needed for the tests, I think it actually looks neater/ more consistent with decimals

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java (hunk 3 above)
Comment (jsotuyod) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java]: No, under the current grammar there is always a prefix, although there may be no suffix: https://github.com/pmd/pmd/blob/master/pmd-java/etc/grammar/Java.jjt#L1830-L1834

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/TccMetric.java (hunk 2 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/TccMetric.java]: Oh, right ;)

**Pull request merged**